### ISSUE TITLE
How to define new distributions?

### ISSUE THREAD

**Issue by allefeld** [scipy/scipy#12133]
I would like to define new distributions using `scipy.stats`. Such a distribution would have a set of parameters, and one could evaluate its PDF, CDF, create random variates etc.

For continuous distributions, there is a short description under "Subclassing" in the documentation of `rv_continuous`. It says one should subclass `rv_continuous`, and as a minimum redefine `_pdf` and `_cdf`, plus possibly a list of others. The description alludes to a lot of things that I don't understand, or understand only very superficially (e.g. "frozen distributions").

I guess one could understand this description if there were some kind of general documentation of the design of `scipy.stats`, but I wasn't able to find it.

Questions:

- Can I redefine `__init__`, and if yes, how should I call `super().__init__`?

- In case I am not supposed to redefine `__init__`, how do I pass parameters to my subclass?

- What are the constraints on the signatures of the methods to be redefined? For example, do I have to provide for distribution parameters to be passed, even though I just want to define them upon instantiation?

- Do I have to support `loc` and `scale`, though I don't want to translate and scale my distribution? The phrase "re-defining at least the _pdf or the _cdf method (normalized to location 0 and scale 1)" seems to suggest that these parameters are automatically taken care of?

- If I implement `_rvs`, can I assume that `self.random_state` is defined? Should I implement a parameter `random_state` myself, or is that taken care of by `rv_continuous.rvs`?

- How does `rvs` pass the number of samples (`size`) to `_rvs`?

- How are the interface methods (e.g. `ppf`) based on the methods I redefine? If I have performance problems with some interface method, which internal method should I additionally redefine? A dependency graph with some indication of performance bottlenecks and loss of precision would be useful.

- > A note on shapes: subclasses need not specify them explicitly. In this case, shapes will be automatically deduced from the signatures of the overridden methods (pdf, cdf etc). If, for some reason, you prefer to avoid relying on introspection, you can specify shapes explicitly as an argument to the instance constructor.

    I have only the faintest idea what this means. How exactly are shapes "automatically deduced from the signatures of the overridden methods"? How can I "specify shapes explicitly as an argument to the instance constructor" if `__init__` is not in the list of methods to be redefined?

These are just the questions I came up with at my current state of understanding, I'm sure if I understood more I had much more to ask.

**Comment by allefeld** [comment]
I now realize that the documentation does not mention `__init__` as one of the methods that can be overridden. I guess the question then becomes, how do I pass parameters to my new class upon instantiation?

And if I can override `__init__`, of course I have to call `super().__init__`. The following additional line at the beginning of `__init__` appears to fix my problem:

    super().__init__(self)

However, `stats.rv_continuous.__init__` has many parameters. Do I have to / should I set some of them? The documentation is sparse:

```
>> stats.rv_continuous.__init__?
Signature:
stats.rv_continuous.__init__(
    self,
    momtype=1,
    a=None,
    b=None,
    xtol=1e-14,
    badvalue=None,
    name=None,
    longname=None,
    shapes=None,
    extradoc=None,
    seed=None,
)
Docstring: Initialize self.  See help(type(self)) for accurate signature.
File:      ~/.conda/envs/std/lib/python3.8/site-packages/scipy/stats/_distn_infrastructure.py
Type:      function
```

**Comment by allefeld** [comment]
> * How does `rvs` pass the number of samples (`size`) to `_rvs`?

By poking around I found that apparently this is done via an instance variable `self._size`? Why would you do it like this (contrary to everything I pretend to know about OOP)? And why isn't it documented anywhere?

**Comment by ilayn** [comment]
Hi @allefeld . I think you are mentioning based on the basic help from http://scipy.github.io/devdocs/tutorial/stats.html
There is also SciPy contributor guide that has a bit more details about the machinery how we add things http://scipy.github.io/devdocs/dev/contributor/adding_new.html#adding-a-new-statistics-distribution

However most of your questions if not all are elaborated in the docstring of `rv_continuous` http://scipy.github.io/devdocs/generated/scipy.stats.rv_continuous.html#scipy.stats.rv_continuous

The best way to get used to it is to see the implementations of different distributions in the source code.

**Comment by ilayn** [comment]
Oops nevermind I just saw you edited the body of the issue

**Comment by allefeld** [comment]
@ilayn, yes, sorry, I made a big mess at first and then tried to create a proper issue.

I would really appreciate some explanation. Thanks for pointing me to the contributor guide! But just by skimming quickly I think that there are still several of my points not covered.

**Comment by ilayn** [comment]
Yes, there is some back and forth needed in the distribution docs but if you go through the methods of `rv_continuous`, for example `fit` would give a bit more info about the shapes, you can familiarize yourself. I am trying not to comment since our domain residents would probably be more helpful, and I would probably confuse you more since I usually butcher my way out when I'm using any statistical tool. 

Maybe you can give a concrete example of a distribution you want to implement so people can explain more carefully? Who knows maybe it would be a candidate to add in the menu we have.

**Comment by josef-pkt** [comment]
Best might be to look at some recent PRs that added new distribution. (I don't remember which right now.)

about `_size` in rvs: this is old behavior #11394 and currently under change 

main point about the structure:
scipy stats distributions are or should be stateless, i.e. each method call need all parameters that define the distribution. There is only one instance of a distribution class, and having statefull attributes causes problems in spilling over to unrelated cases.
The disadvantage is that we cannot store the results of intermediate computations which would speed up the method calls.

subclasses should only redefine the methods that start with underline, e.g. `_pdf`, (and maybe `fit`). The inherited `pdf`, `ppf`, ... handle loc-scale, broad casting, domain checking, and have generic version for methods that are not defined in the specific distribution class.

**Comment by josef-pkt** [comment]
I went through my email and got confused about the mixture distribution case.

If you are NOT writing a new distribution class for scipy.stats and only want to use it for your own usecase or library, then there is no restriction on not having statefull attributes.
But then, each instance of the class will be specific to the arguments or attributes define in `__init__`, and you shouldn't create a single instance as does scipy.stats.distributions

two examples from statsmodels

zero inflated poisson follows the scipy.stats convention, it hard codes the distribution, and uses all parameter as method arguments, no `__init__`  
(aside `rvs` is slow because it doesn't define `_rvs` and uses the slow generic method.)
https://github.com/statsmodels/statsmodels/blob/master/statsmodels/distributions/discrete.py#L45

ancient code of mine that I haven't looked at in a long time:
(The code has some minimal unit tests, so it should still be working after 10 to 12 years)

Define distribution and other things in `__init__` which don't show up as method arguments
https://github.com/statsmodels/statsmodels/blob/master/statsmodels/sandbox/distributions/transformed.py#L55

If we want a single instance, then it has to define all extra `__init__` arguments, and is then specific to that case. The remaining parameters for the distribution are don't define a state through attributes
https://github.com/statsmodels/statsmodels/blob/master/statsmodels/sandbox/distributions/transformed.py#L115

**Comment by allefeld** [comment]
@josef-pkt, thanks for the pointers!

I wasn't aware that there are only single instances of distribution classes. (Is this documented anywhere?) I admit the fact that the distribution parameters can or should be included in each method call already indicates that.

What I don't understand is why you chose to do it this way? You already mentioned performance issues, but even apart from that, from my (maybe naive) perspective it appears to be just plain illogical.

IMHO the OOP logic maps nicely on the logic of distributions. A parametrized family of distributions =  a class, a specific distribution with fixed parameters = an instance. If one family is a special case of another, this can be modeled by inheritance. There can be different instances with the same parameters, they model i.i.d. RVs.

I also don't understand why you include location and scale parameters for every distribution. In some cases that makes sense, e.g. relative frequencies are scaled-binomial, but in many others it doesn't. Apparently it also makes it hard to specify parameters for some distributions, see [A rant on scipy.stats](https://clementc.github.io/blog/2017/06/15/rant_on_scipy_stats/). This problem has even given rise to a wrapper library, [distcan](https://github.com/sglyon/distcan). I would have implemented optional translation and scaling by a wrapper class, e.g. `LocationScale(dist, loc, scale)`, where `dist` can be an arbitrary other distribution.

I realize criticism is cheap and you put a lot of work into this. But I have to say these for me strange design decisions have already impeded my use of the library, and right now make it very hard to build on it.

***

As for my own implementation, immediately it is only meant for my own use. Here is my current version, and it seems to work fine so far:

```
class Mixture (stats.rv_continuous):
    "Mixture distribution"
    
    def __init__(self, distributions, probabilities):
        super().__init__(self)
        self.dists = distributions
        probabilities = np.array(probabilities)
        self.probs = probabilities / sum(probabilities)

    def _pdf(self, x):
        pdf = 0
        for dist, prob in zip(self.dists, self.probs):
            pdf = pdf + prob * dist.pdf(x)
        return pdf

    def _cdf(self, x):
        cdf = 0
        for dist, prob in zip(self.dists, self.probs):
            cdf = cdf + prob * dist.cdf(x)
        return cdf
    
    def _rvs(self):
        N = np.prod(self._size)
        ns = np.random.multinomial(N, self.probs)
        samples = []
        for dist, n in zip(self.dists, ns):
            samples.append(dist.rvs(size=n))
        samples = np.hstack(samples)
        samples = np.random.permutation(samples).reshape(self._size)
        return samples
```

**Comment by rgommers** [comment]
Wow, I had never heard of `distcan`. That looks potentially very useful. I don't think there's been a consistent proposal before on folding something like that into `scipy.stats`.

**Comment by allefeld** [comment]
I'm not sure that this is interesting to you, but just in case, here is an example of a tiny class hierarchy modeling distributions that I defined (outside of the `scipy.stats` framework, but with a similar interface). I think it demonstrates the advantage of passing parameters to `__init__`. The point of the `Empirical` class is that its CDF is the ECDF, and it can be used to create bootstrap replications.

Just to be clear: What I do like about your approach is that you have inherited default implementations based e.g. on numerical integration, which use private methods if present.

```
import numpy as np

class Discrete:
    """
    discrete distribution
    """
    
    def __init__(self, values, weights=None):
        # input checking and normalization
        values = np.array(values).flatten()
        values, index, counts = np.unique(values, return_index=True, return_counts=True)
        if np.any(counts > 1):
            raise ValueError('values must be unique')
        if weights is not None:
            weights = np.array(weights).flatten()
            weights = weights[index]
            if len(weights) != len(values):
                raise ValueError('number of weights must match number of values')
        else:
            weights = np.ones(len(values))
        # set state
        self.xk = values
        self.pk = weights / sum(weights)
        self.Pk = np.cumsum(self.pk)
        self.Pk0 = np.hstack((0, self.Pk)) # for CDF lookup
        
    def pmf(self, x, tol=0.0):
        """
        probability mass function
        """
        # normalization
        x = np.array(x)
        shape = x.shape
        x = x.flatten()
        # compute
        e = np.abs(x - self.xk[:, np.newaxis])
        ind = np.argmin(e, axis=0)
        p = self.pk[ind]
        eMin = e[ind, np.arange(e.shape[1])]
        p[eMin > tol] = 0.0
        # back into shape
        p = np.reshape(p, shape)
        return p
    
    def cdf(self, x):
        """
        cumulative distribution function
        """
        # normalization
        x = np.array(x)
        shape = x.shape
        x = x.flatten()
        # compute
        ind = np.sum(self.xk[:, np.newaxis] <= x, axis=0)
        P = self.Pk0[ind]
        # back into shape
        P = np.reshape(P, shape)
        return P
    
    def __call__(self, x):
        """
        F(x) is a shortcut for F.cdf(x).
        """
        return self.cdf(x)
    
    def rvs(self, size=None, rng=None):
        rng = np.random.default_rng(rng)
        return rng.choice(self.xk, p=self.pk, size=size)

    def support(self):
        return self.xk
    
    def __repr__(self):
        return f'Discrete({repr(self.xk)}, {repr(self.pk)})'

    
class Empirical(Discrete):
    """
    empirical distribution
    """

    def __init__(self, sample):
        (xu, cu) = np.unique(sample, return_counts=True)
        super().__init__(xu, cu)
        self.n = len(sample)
    
    def bootstrap_sample(self, N=1, rng=None):
        rng = np.random.default_rng(rng)
        if N == 1:
            return self.rvs(size=self.n, rng=rng)
        else:
            return self.rvs(size=(N, self.n), rng=rng)
```

**Comment by josef-pkt** [comment]
I agree, the single instance is because of the old aversion to OOP. 
But loc-scale families and having a pretty strict pattern in distribution also has the advantage that we have a single framework that can implement a lot of features with little effort because distributions don't have to be special cased. 

We are getting more proposals for distributions that don't fit well in the current pattern.
I don't see a problem adding new classes outside of the current framework, with maybe reusing parts that are still common.

However, the current pattern works very well for most standard distributions, e.g. almost all my pvalue calculation are just norm.sf(...), t.sf(..), f.sf(..) and some noncentral versions for power.

**Comment by josef-pkt** [comment]
distcan and similar package before that don't work for implementing distributions in a generic way.

It's built on too many special cases, and those special cases are often not even the same across fields. On Wikipedia several distributions have several "canonical" parameterizations. For other distributions, there are different versions depending on whether they include loc and/or scale, i.e. in the style of 2-parameter pareto vs 3 parameter pareto, and similarly for many distributions.
(some R packages have different distribution names depending on the number of parameters including loc, scale)

statsmodels GLM uses the mean-variance parameterization which is standard in the GLM literature but not "canonical" in the distribution literature. To work around we have functions to convert parameters from GLM to scipy (with occassional problems in figuring out the general parameter transformation. https://stackoverflow.com/questions/60215085/calculating-scale-dispersion-of-gamma-glm-using-statsmodels )

IMO it's better to have a somewhat restrictive framework with 100 distributions and occassional hiccups than to have 15 user friendly distributions with an interface that is "canonical" for users but not maintainers.

**Comment by allefeld** [comment]
@rkern, thanks for the detailed explanation!

As someone coming from Matlab, the intention to avoid OOP sounds strange to me. One of the reasons to switch is that Python is a much more expressive language than Matlab. The old statistics interface in Matlab was indeed a set of functions, but standalone, not bundled in classes or instances. And the modern one in the Stats & ML Toolbox does use classes & objects, unless I'm mistaken in pretty much the way I sketched. Don't get me wrong, I don't think that OOP should be forced upon everything, but I think it is a natural match to modelling distributions.

I don't think that `loc` and `scale` should be removed, it just shouldn't be the main way to specify parameters, except for explicit location-scale families. You are right, sometimes applications come up that call for e.g. scaling even though the standard definition does not allow for that. The concern is more that focusing on these parameters may make it harder to create a distribution from "canonical" parameters.

@josef-pkt you are right that there are many cases in which there is not *one* set of canonical parameters. The normal distribution with standard deviation vs variance is one example, and Wikipedia notes three different parametrizations for gamma. In my ideal world sketched above, there could be several "factory" class methods that transform parameters into the format used by `__init__`, and maybe get methods that transform internal parameters back.

I don't think that `distcan` can replace (the interface of) `scipy.stats`; yes, it is much smaller. But I think more like @rgommers that this is something that could be integrated into `scipy.stats`, making it more user-friendly (you do want it to be user-friendly, right?) without messing with internals.

But @josef-pkt, I see a bit of a contradiction between criticising `distcan` for supporting only 15 distributions as opposed to the 100 of `scipy.stats`, but at the same time defending `loc`/`scale` with the argument that the pattern works for "norm.sf(...), t.sf(..), f.sf(..) and some noncentral versions for power". Also, noncentral t is *not* just a shifted version of t.

**Comment by josef-pkt** [comment]
The basic structure of the distributions is 10 to 15 years old, AFAIK, and never has fundamentally change. At that time Python was still the underdog, and many developers thought the API has to cater to matlab conventions.

I'm not really criticising distcan or similar. I think they contribute a good user interface and IMO scipy should eventually add more support for less confusing interfaces or distributions that don't fit the standard pattern easily. For example, something like frozen distributions could also transform parameters instead of just storing them. But that needs distribution specific work.

However, scipy needs to implement a large number of distribution, also so that other packages can delegate to it. A new distribution that fits the current pattern can be implemented in 15 lines or so and gains all the general features, extensions and even unit tests. 

(The same is true for me in statsmodels. For years I have been searching for generic solutions that makes it possible to easily reuse features for new models or to add features that immediately apply to many existing models. This reduces developer time and makes the code more robust because unit tests evaluate features for many different use cases. That doesn't always create an "obvious" api, but then there can be wrappers outside or inside the package.
In short: I'd rather have users complaining that something is not convenient, than complaining that things are not available in Python at all.)

norm.sf(...), t.sf(..), f.sf(..) were more examples where having a single instance is convenient.
stats.TDist().sf(...)   doesn't look as pretty as stats.t.sf(...)

**Comment by allefeld** [comment]
Alright, I didn't expect that you would overhaul the API of `scipy.stats` because of my issue or someone else's rant. Just take it as my 2 cents. Thanks for engaging with me on that.

To come back to my original question: I take it there is currently not one guide to defining new distributions that covers everything one needs to know (at least none of those I was pointed to cover everything I asked). The advice is basically to go into the code and see how it is done? So maybe I can make **this request**: That someone who knows this stuff in and out puts such a complete guide together.

I think that would be useful not just for aspiring contributors, but also for simple users. At least in my work, I frequently feel the need to define a particular distribution that isn't useful for anyone else (or at least not many people). And then it is very useful not to have to define everything explicitly, but to benefit from the `scipy.stats`' automatic filling in, e.g. getting `ppf` for the price of `cdf`.

And **can I ask**, is my implementaton `Mixture` above correct for a distribution that is not to be contributed?

**Comment by WarrenWeckesser** [comment]
> So maybe I can make this request...

You might be interested in my email to the scipy-dev mailing list on May 8: https://mail.python.org/pipermail/scipy-dev/2020-May/024157.html

**Comment by josef-pkt** [comment]
Your Mixture class looks fine to me.

I don't know what the current status is with names and similar. They were some required items like adding `name` of distribution and number or names of shapes.
Shape parameters and name information will not be relevant if there are no parameters.

Your mixture class doesn't have any free parameters. If you need to change a parameter for the underlying distribution, then you need to create a new instance.

If you have bounded support, then I think you still need to implement the limits `.a` `.b`.

**Comment by allefeld** [comment]
@WarrenWeckesser, that looks promising!

@josef-pkt, thanks.

**Comment by pvanmulbregt** [comment]
If one's viewpoint is that distributions are really a family of distributions, and the main use is to determine the shape parameters (e.g. via `fit()`), or to evaluate the distribution's `pdf|cdf|sf|...` for many shape parameters simultaneously at one `x` value (or of `x` broadcasted with the shape parameters) then a single instance may fit reasonably well.

OTOH, if one's viewpoint is that a distribution really is just one member of a family, specified by the shape parameters, then a single instance can be quite kludgy to deal with. 

Some distributions can benefit computationally by precomputing some data (e.g. `truncnorm` which benefits from calculating the probability in the truncated region as few times as possible.)  Placing that precomputed data inside the single instance breaks thread-safety. This can be worked around but takes a little effort, and depends on the assumption that the `pdf|cdf|..` are being called with many `x` values at once, so the overhead is amortized.

The fact that the frozen distributions are just the pair `(distribution, shape parameters)` without any ability to store precomputed data can be confusing to users.

One big drawback of the single instance approach is the difficulty of dealing with non-numeric shape parameters.  E.g. gh-12000 which attempts to add the Poisson binomial distribution.  The single shape parameter is really an arbitrary length list of probabilities. There is substantial overhead to computing the `pmf|cdf`.  Broadcasting can get in the way also.
Similarly gh-6800 was about adding mixtures of distributions, and doesn't fit so easily into the current framework

**Comment by allefeld** [comment]
I guess where specifying distribution parameters in each call comes in handy is when one wants to evaluate e.g. the pdf for many parameter values at once, and broadcasting can be used. Creating one instance for each set of parameters would be inefficient in that case.

@pvanmulbregt how does placing precomputed data inside the single instance break thread-safety?

**Comment by josef-pkt** [comment]
> how does placing precomputed data inside the single instance break thread-safety?

Two threads might access and change the one instance at the same time.

When I initially fixed the distribution code (10 to 12 years ago), there were many cases where attributes were changed by one call but not cleaned up and it produced incorrect results on the next call with different arguments.

Stale state spillovers are a lot of fun to debug, especially when they only show up some of the time.

**Comment by allefeld** [comment]
> Two threads might access and change the one instance at the same time.

Couldn't that be prevented by setting instance data only in `__init__`, and not allowing modifications later?

**Comment by josef-pkt** [comment]
> Couldn't that be prevented by setting instance data only in __init__, and not allowing modifications later?

That works for attributes that don't depend on the arguments/parameters of the distribution.
For example, the support bounds `.a`, `.b` in Poisson and similar are fixed. For cases where .a and .b depend on the parameters, it is more difficult because then they are also attributes that are not allowed to spill over  and need to be set in each call.

**Comment by pvanmulbregt** [comment]
In particular, the distribution's `__init__` is called when `stats.py`is loaded, not the first time the distribution is used.

**Comment by mdhaber** [comment]
@allefeld does this answer your questions?

**Comment by allefeld** [comment]
@mdhaber, well I certainly got a lot of information. But I asked 8 questions with a hint there might be more to follow, and only some of them have been discussed.

More importantly, I still think one shouldn't have to go through source code or issues in order to find out these details.
There should be a page in the documentation that includes all the necessary information needed for a *user* to define a new distributions for their own use, so that the respective object can be used in place of a `scipy.stats´ distribution.

And there could also be a second page that explains how to construct it such that it can be contributed to the package itself (developer's guide).

**Comment by mdhaber** [comment]
I agree. It looks like @WarrenWeckesser's is starting to work on these things. His first PR towards this end is gh-12069. It may look more developer-oriented than you had in mind, but I agree that it would be a good idea to make it more clear for users. I recently started working with stats, too, and I also found it confusing. 
In the meantime, if you do get the answers to your questions and would like to share some of your hard-earned lessons, it would be great if you could draft the document you wish you had when you started! Perhaps the experts could help polish it, but the best perspective for that sort of thing is from someone who recently learned everything the hard way. The [wiki](https://github.com/scipy/scipy/wiki) might be a good place to start before worrying about rendering with Sphinx.
Thanks for your patience. Almost everything is done on a volunteer basis, so this might take a bit of time.

**Comment by Evidlo** [comment]
I was interested in adding a `__repr__` and some other minor customizations to `scipy.stats.norm`, but my first few guesses  were not successful.

Attempt 1: without reading any docs.  I just assumed `scipy.stats.norm` was a plain-old class and e.g. `.pdf` was a classmethod with some logic to check whether the RV was frozen or not.

```
class Gaussian(scipy.stats.norm):
    def __repr__(self):
        return f'N({self.mean()}, {self.var()})'
belief = Gaussian(0, 1)
print(belief)

>>> <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7f13e2af4190>
```

Attempt 2: after reading the [docs on rv_continuous](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.html)

```
class Gaussian(scipy.stats.rv_continuous):
    def __repr__(self):
        return f'N({self.mean()}, {self.var()})'

    _pdf = scipy.stats.norm._pdf
belief = Gaussian(0, 1)
print(belief)

>>> N(0.24197072451914342, 0.3420761469262813)
```
